test-data/unit/check-incremental.test

```diff
@@ -3376,3 +3376,65 @@ def f() -> None:
 [out2]
 [out3]
 [out4]
+
+[case testTreeShadowingViaParentPackage]
```

Comment:
I think that's because in master *no* trees are preserved -- that's #4198, and it completely masks #4208. I suppose I could split the PR in two parts, the part that fixes #4198, at which point #4208 becomes apparent, and then the part that fixes the latter. But that's a fair bit of work. I do know that at some point in the past (I think after the first dmypy PR) I saw the same errors as #4208, and when I was three quarters through this PR I saw those same errors. I am confident that this test demonstrates the specific issue that was fixed.